starboard/nplb/posix_compliance/posix_environment_test.cc

```diff
@@ -0,0 +1,443 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <unistd.h>
+
+#include <algorithm>
+#include <cerrno>
+#include <cstring>
+#include <iterator>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+// This test is placed first to verify the state of 'environ' before
+// any other tests (especially those using fixtures with SetUp/TearDown
+// that might call setenv/getenv/unsetenv) are run.
+TEST(PosixEnvironmentGlobalStateTests, EnvironIsAccessibleAtStartup) {
+  ASSERT_NE(nullptr, environ) << "The global 'environ' variable should not be "
+                                 "nullptr at program startup";
+}
+
+// Assert that environ is declared and accessible, and that its type is
+// 'char**'.
+static_assert(std::is_same_v<decltype(environ), char**>,
+              "'environ' is not declared or is not of type 'char**'");
+
+// TODO: b/390675141 - Remove this after non-hermetic linux build is removed.
+// On non-hermetic builds, getenv() and setenv() are declared "noexcept".
+#if BUILDFLAG(IS_COBALT_HERMETIC_BUILD)
+// Assert that getenv has the signature: char* getenv(const char*)
+static_assert(std::is_same_v<decltype(getenv), char*(const char*)>,
+              "'getenv' is not declared or does not have the signature "
+              "'char* (const char*)'");
+
+// Assert that setenv has the signature:
+// int setenv(const char*, const char*, int)
+static_assert(
+    std::is_same_v<decltype(setenv), int(const char*, const char*, int)>,
+    "'setenv' is not declared or does not have the signature "
+    "'int (const char*, const char*, int)'");
+#endif
+
+// This test fixture base class is designed to be inherited by the other test
+// harnesses. It captures a snapshot of the environment variables
+// at the beginning of SetUp and verifies that the environment is identical
+// at the end of TearDown. This ensures that tests do not have side effects
+// on the process environment.
+class EnvironmentTestBase : public ::testing::Test {
+ public:
+  EnvironmentTestBase() {
+    // This runs BEFORE any SetUp() method.
+    // We snapshot the initial environment state.
+    for (char** env = environ; *env != nullptr; ++env) {
+      initial_environ_.push_back(*env);
+    }
+    std::sort(initial_environ_.begin(), initial_environ_.end());
+  }
+
+  ~EnvironmentTestBase() override {
+    // This runs AFTER any TearDown() method.
+    // We take a final snapshot and verify it against the initial one.
+    if (!HasFatalFailure()) {
+      std::vector<std::string> final_environ;
+      for (char** env = environ; *env != nullptr; ++env) {
+        final_environ.push_back(*env);
+      }
+      std::sort(final_environ.begin(), final_environ.end());
+
+      EXPECT_EQ(initial_environ_.size(), final_environ.size())
```

Comment:
Are there any potential issues with running EXPECT after TearDown occurs or is it valid gtest practice?